firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/I2C/HoneywellABP/Sensor.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2020, the Pez Globo team and the Pufferfish project contributors
+ *
+ *      Author: Rohan Purohit
+ *
+ */
+
+#include "Pufferfish/Driver/I2C/HoneywellABP/Sensor.h"
+
+namespace Pufferfish::Driver::I2C::HoneywellABP {
+
+// StateMachine
+
+StateMachine::Action StateMachine::update() {
+  switch (next_action_) {
+    case Action::initialize:
+      next_action_ = Action::measure;
+      break;
+    case Action::measure:
+      break;
+  }
+  return next_action_;
+}
+
+// Sensor
+
+InitializableState Sensor::setup() {
+  switch (next_action_) {
+    case Action::initialize:
+      return initialize();
+    case Action::measure:
+      return InitializableState::ok;
+  }
+  return InitializableState::failed;
+}
+
+InitializableState Sensor::output(float &output) {
+  switch (next_action_) {
+    case Action::measure:
+      return measure(output);
+    default:
+      break;
+  }
+  return InitializableState::failed;
+}
+
+InitializableState Sensor::initialize() {
+  if (retry_count_ > max_retries_setup) {
+    return InitializableState::failed;
+  }
+
+  retry_count_ = 0;
+  // Test
+  while (device_.test() != I2CDeviceStatus::ok) {
```

Comment:
Let's refactor the `test` method out of the `Device` class; test functionality to be carried out during initialization should instead be implemented in the `Sensor::initialize()` method, or else in private methods called by `Sensor::initialize()`.